Shoot at a speed

wmarshall opened the pull request:
Submitting PR for Code Review per https://trello.com/c/IQ4JpBgu/13-shoot-at-a-speed

Commit:
Created Shooter subsystem, commands, and associated buttons and constants

-Currently, there is an encoder for speed tracking purposes; there are no control loops set up for this as of yet
-Commands have been attached to associated buttons, and code should be ready for use once TODOs are resolved

Commit:
Created p-loop methods for controlling shooter speed

-Along with relevant constants
-Throttle also added for Shooter speed control
-Comments to be added
-Code is currently being checked by mentors

Ruby18 committed:
Made changes to Shooter according to the code review.

wmarshall commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -62,4 +67,35 @@ public static double getRightY() {
+
+	public OI(){
+		shootButt.whileHeld(shooterOn);
+		stopShootButt.whenPressed(shooterStop);
```
How would this control be used if the default behavior is to stop shooting when the button is released?

Num3ra1s replied:
The stop shoot button is a safety button. If you think it's unnecessary, we can take it out.

wmarshall commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -1,6 +1,11 @@
 package org.usfirst.frc.team2399.robot;
 
+import org.usfirst.frc.team2399.robot.commands.JoyShoot;
+import org.usfirst.frc.team2399.robot.commands.ShooterStop;
```
These classes, JoyShoot and ShooterStop, are not named consistently

Num3ra1s replied:
It was Lauren's executive decision to rename these after merge, we will establish naming conventions in a journal tonight.

wmarshall commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -62,4 +67,35 @@ public static double getRightY() {
+	 * Throttle Methods
+	 * (throttle +1)/2 sets the throttle range from 0 to 1 rather than
+	 * 	-1 to 1
+	 * TODO: Figure out which joystick we actually want to get this value from
```
Should be resolved before merge

wmarshall commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -16,12 +16,17 @@
 	 * Address for Talons
 	 * Things on the CAN network have addresses. Things on PWM network have
 	 * ports.
+	 * TODO: Set actual address
```
Should be resolved before merge

wmarshall commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -32,12 +37,51 @@
 	/**
 	 * Forward Constants
 	 * To ensure positive values of motion are what you expect.
+	 * TODO: Set actual constants
```
Resolve before merge

wmarshall commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -32,12 +37,51 @@
+	 */
+	public static final int SHOOTER_ENCODER_COUNT = 920348039;
+	public static final double SHOOTER_GEAR_RATIO = 34023984;
+	public static final double SHOOTER_WHEEL_CIRCUMFERENCE = 203948;
```
Should be non-garbage numbers before testing. Ask design.

wmarshall commented on src/org/usfirst/frc/team2399/robot/commands/JoyShoot.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class JoyShoot extends Command {
+	
+	double speed;
```
Should this be private?

Num3ra1s replied:
We're not sure why - all our commands from last year were public, and we need access to them in order to pair commands with buttons. If you could explain why they should be private, or point us in the right direction toward that explanation, that would be great!

wmarshall commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -32,12 +37,51 @@
+	public static final int SHOOTER_ENCODER_COUNT = 920348039;
+	public static final double SHOOTER_GEAR_RATIO = 34023984;
+	public static final double SHOOTER_WHEEL_CIRCUMFERENCE = 203948;
+	public static final double SHOOTER_DISTANCE_PER_PULSE =  SHOOTER_WHEEL_CIRCUMFERENCE
```
What is this value used for?

wmarshall commented on src/org/usfirst/frc/team2399/robot/commands/ShooterStop.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class ShooterStop extends Command {
+	
+	private Shooter shooter = Robot.shooter;
```
Indentation issues

Num3ra1s replied:
There aren't issues on Chloe's computer as of now. We're not sure what happened here.

wmarshall commented on src/org/usfirst/frc/team2399/robot/commands/JoyShoot.java:
```diff
@@ -0,0 +1,49 @@
+
+    // Make this return true when this Command no longer needs to run execute()
+    protected boolean isFinished() {
+        return false;
```
What makes this command stop?

Num3ra1s replied:
Nothing makes this stop currently. Is there a reason to stop joystick control of the shooter once teleop has begun?

wmarshall commented on src/org/usfirst/frc/team2399/robot/commands/ShooterStop.java:
```diff
@@ -0,0 +1,45 @@
+
+    public ShooterStop() {
+    	requires(shooter);
+    	setInterruptible(true);
```
Good, can be interrupted by command to shoot.

wmarshall commented on src/org/usfirst/frc/team2399/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+		shooterTalon.setFeedbackDevice(FeedbackDevice.CtreMagEncoder_Relative);
+		shooterTalon.changeControlMode(TalonControlMode.Speed);
+		shooterTalon.reverseOutput(RobotMap.SHOOTER_FORWARD_CONSTANT == -1);
+		shooterTalon.reverseSensor(false);
```
Don't we have a sensor reverse constant?

wmarshall commented on src/org/usfirst/frc/team2399/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+	// Sets the speed to the inputted speed * the forward constant
+	// Position change/10 ms
+	public void setShooterSpeed(double speed) {
+		shooterTalon.set(speed * RobotMap.SHOOTER_FORWARD_CONSTANT);
```
Why are we multiplying by SHOOTER_FORWARD again? Doesn't CANTalon.reverseOutput() do this?

wmarshall commented on src/org/usfirst/frc/team2399/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+	 *
+	 */
+
+	// Gets speed from sensor in ticks/100 ms
```
Javadoc comment?

wmarshall commented on src/org/usfirst/frc/team2399/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+// Allows us to Increment and Decrement for accuracy 
+	public void incrementSpeedConstant() {
+		double currentPConstant = shooterTalon.getP();
+		shooterTalon.setP(currentPConstant += RobotMap.SHOOTER_SPEED_INCREMENT_CONSTANT);
```
Why are we doing this AND shooterTalon.getP()?

Num3ra1s replied:
This was separated into two lines (with the current p value assigned to a variable) because it was easier to determine the function of the method when it was written that way. We can condense it to one line, but it's harder for us to determine the function of the method when it's written like that.

wmarshall commented on src/org/usfirst/frc/team2399/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+
+	// Default is the motor at 0 speed
+	public void initDefaultCommand() {
+		setDefaultCommand(new ShooterStop());
```
Good, nice safe default

wmarshall commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -32,12 +37,51 @@
+	 * P-Loop Constants
+	 */
+	public static final int SHOOTER_SPEED_P_CONSTANT = 0;
+	public static final int SHOOTER_HERTZ_CONSTANT = 0;
```
Do we still need this constant?

wmarshall reviewed (commented):
No pass. See comments

wmarshall commented on src/org/usfirst/frc/team2399/robot/commands/ShooterStop.java:
```diff
@@ -0,0 +1,45 @@
+    //When the command is run, the shooter is set to this speed
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    	shooter.setShooterSpeed(0);
```
Don't we have a stop constant?

Commit:
Code review for shooter implemented

-Some TODOs still unresolved due to lack of information (to be fixed later tonight/tomorrow or as soon as information is available)

Commit:
Further code review issues resolved

Commit:
More review changes implemented

Commit:
Comments for Shooter updated

wmarshall commented:
Passed Review/good to test on 2/1

Commit:
Merge branch 'master' into Shooter

# Conflicts:
#	src/org/usfirst/frc/team2399/robot/OI.java
#	src/org/usfirst/frc/team2399/robot/Robot.java
#	src/org/usfirst/frc/team2399/robot/RobotMap.java

chloe6248 committed:
Updated .classpath

chloe6248 committed:
Updated .classpath
